zircon.core/src/commonMain/kotlin/org/hexworks/zircon/api/component/builder/ComponentBuilder.kt

```diff
@@ -14,63 +20,175 @@ import org.hexworks.zircon.api.grid.TileGrid
+     */
+    val size: Size
+
+    fun withName(name: String): U {
+        this.name = name
+        return this as U
+    }
+
     /**
      * Sets if the [Component] should be updated when it is attached to a parent
      * or not. By default the [Component]'s common properties ([ComponentProperties])
      * will be updated from its parent.
      */
-    fun withUpdateOnAttach(updateOnAttach: Boolean): U
+    fun withUpdateOnAttach(updateOnAttach: Boolean): U {
+        this.updateOnAttach = updateOnAttach
+        return this as U
+    }
 
     /**
      * Sets the [ComponentStyleSet] the [Component] will use.
      */
-    fun withComponentStyleSet(componentStyleSet: ComponentStyleSet): U
+    fun withComponentStyleSet(componentStyleSet: ComponentStyleSet): U {
+        this.componentStyleSet = componentStyleSet
+        return this as U
+    }
 
     /**
      * Sets the [Tileset] to use for the [Component].
      */
-    fun withTileset(tileset: TilesetResource): U
+    fun withTileset(tileset: TilesetResource): U {
+        this.tileset = tileset
+        return this as U
+    }
 
     /**
      * Sets the [ColorTheme] to use for the [Component].
      */
-    fun withColorTheme(colorTheme: ColorTheme): U
+    fun withColorTheme(colorTheme: ColorTheme): U {
+        this.colorTheme = colorTheme
+        return this as U
+    }
 
     /**
      * Sets the [AlignmentStrategy] to use for the [Component].
      */
-    fun withAlignment(alignmentStrategy: AlignmentStrategy): U
+    fun withAlignment(alignmentStrategy: AlignmentStrategy): U {
+        this.alignment = alignmentStrategy
+        return this as U
+    }
 
     /**
      * Sets the [ComponentDecorationRenderer]s for the resulting [Component].
      * The component will be decorated with the given decorators in the given
      * order.
      */
-    fun withDecorations(vararg renderers: ComponentDecorationRenderer): U
+    fun withDecorations(vararg renderers: ComponentDecorationRenderer): U {
+        this.decorations = renderers.asList()
+        return this as U
+    }
 
     /**
      * Sets the [ComponentRenderer] for the resulting [Component].
      */
-    fun withComponentRenderer(componentRenderer: ComponentRenderer<T>): U
+    fun withComponentRenderer(componentRenderer: ComponentRenderer<T>): U {
+        this.componentRenderer = componentRenderer
+        return this as U
+    }
 
     /**
      * Creates a [ComponentRenderer] for the resulting [Component] using the
      * given component renderer [fn].
      */
-    fun withRendererFunction(fn: (TileGraphics, ComponentRenderContext<T>) -> Unit): U
+    fun withRendererFunction(fn: (TileGraphics, ComponentRenderContext<T>) -> Unit): U {
+        this.renderFunction = fn
+        return this as U
+    }
+
+    /**
+     * Adds component(s) to the target ([T]) [Component] using the
+     * component builder DSL.
+     */
+    fun withComponents(fn: U.() -> Unit): U {
+        fn(this as U)
+        return this
```

Comment:
I'm not sure what the use case for this is, as to me it looks identical to `.apply {}` in syntax. All it does is open its own scope, and since it's visible to the DSL you can end up with funny looking code like
```
    buildHbox {
        // HBoxBuilder scope
        withComponents {
            // same HBox builder scope
        }
    }
```

Reply:
That's a good idea. I'll remove this.